cedar-lean/Cedar/Thm/Validation.lean

```diff
@@ -0,0 +1,70 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec
+import Cedar.Data
+import Cedar.Validation
+import Cedar.Thm.Validation.Validator
+
+/-!
+This file contains the top-level correctness properties for the Cedar validator.
+We show that if validation succeeds, then for any request consistent with the schema,
+evaluation of every policy is "correct", that is, it either produces a boolean value,
+or throws one of the errors in the set of valid errors.
+--/
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+/--
+If a set of policies is well-typed (valid) with respect to the schema according to the validator,
+and the input request and entities are consistent with the schema, then every policy "evaluates correctly":
+this means that evaluation either produces a boolean value, or throws an error of type
+`entityDoesNotExist`, `extensionError`, or `arithBoundsError`. These errors cannot be protected against at
+validation time, as they depend on runtime information. 
+-/
+theorem validation_is_sound (policies : Policies) (schema : Schema) (request : Request) (entities : Entities) :
+  validate policies schema = .ok () →
+  RequestAndEntitiesMatchSchema schema request entities →
+  AllEvaluateCorrectly policies request entities
+:= by
+  intro h₀ h₁
+  unfold validate at h₀
+  simp only [AllEvaluateCorrectly]
+  cases h₃ : policies with
+  | nil => simp only [List.not_mem_nil, false_implies, implies_true]
+  | cons h' t' =>
+    intro policy pin
+    simp only [OneEvaluatesCorrectly]
+    apply typecheck_policy_with_environments_is_sound policy schema.toEnvironments request entities h₁
+    subst h₃
+    simp only [List.forM_cons'] at h₀
+    cases h₄ : (typecheckPolicyWithEnvironments h' schema.toEnvironments) <;> simp only [h₄,
+      Except.bind_err] at h₀
```

Comment:
Nit:   for readability, can we putt the entire `simp only ...` step on line 58, or, alternatively, break the line after `<;>`?